File: 383/main.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        """
        approach
        count letter frequencies
        """
        count = {}
        for c in magazine:
            if c not in count:
                count[c] = 0
            count[c] += 1
        for c in ransomNote:
            if c not in count:
                return False
            if count[c] == 0:
                return False
            count[c] -= 1
        return True

File: 383/main.ts
function canConstruct(ransomNote: string, magazine: string): boolean {
  let counts = new Map();
  for (let char of magazine) {
    if (!counts.has(char)) {
      counts.set(char, 0);
    }
    counts.set(char, counts.get(char) + 1);
  }
  for (let char of ransomNote) {
    if (!counts.has(char)) {
      return false;
    }
    if (counts.get(char) <= 0) {
      return false;
    }
    counts.set(char, counts.get(char) - 1);
  }
  return true;
}
